From 84437a039edc701636c20a117bca542185758ef2 Mon Sep 17 00:00:00 2001
From: Joshua Ashton <joshua@froggi.es>
Date: Fri, 7 May 2021 20:30:34 +0100
Subject: [PATCH 1/3] [dxgi] Add global factory for RE8

This game spams CreateDXGIFactory calls like there's no tomorrow. Cache this.
---
 src/dxgi/dxgi_main.cpp | 17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/src/dxgi/dxgi_main.cpp b/src/dxgi/dxgi_main.cpp
index 17530f53e..98e977772 100644
--- a/src/dxgi/dxgi_main.cpp
+++ b/src/dxgi/dxgi_main.cpp
@@ -1,11 +1,26 @@
 #include "dxgi_factory.h"
 #include "dxgi_include.h"
 
+#include "../util/util_env.h"
+
 namespace dxvk {
   
   Logger Logger::s_instance("dxgi.log");
-  
+
+  bool useGlobalFactory() {
+    static bool s_useGlobalFactory = env::getExeName() == "re8.exe";
+    return s_useGlobalFactory;
+  }
+
+  HRESULT getGlobalFactory(REFIID riid, void **ppFactory) {
+    static Com<DxgiFactory, false> s_factory = new DxgiFactory(0);
+    return s_factory->QueryInterface(riid, ppFactory);
+  }
+
   HRESULT createDxgiFactory(UINT Flags, REFIID riid, void **ppFactory) {
+    if (useGlobalFactory())
+      return getGlobalFactory(riid, ppFactory);
+
     try {
       Com<DxgiFactory> factory = new DxgiFactory(Flags);
       HRESULT hr = factory->QueryInterface(riid, ppFactory);
